Curves: Added serialization/ deserialization

Mugen87 opened the pull request:
This PR adds serialization/ deserialization for all curve entities. It's now possible to store and load `ShapeGeometry` and `ShapeBufferGeometry`.

The implementation allows to perform serialization/ deserialization on all hierarchy levels (`Curve`, `CurvePath`, `Path`, `Shape`) with `.toJSON()` and `.fromJSON()` methods.

I'll consider `ExtrudeGeometry` in a separate PR.

Demo: https://jsfiddle.net/ht3zuzkr/2/

Mugen87 committed:
Curves: Added serialization/ deserialization

Mugen87 committed:
Curves: Init arrays before deserialize

Mugen87 commented:
This PR directly addresses #8007

Mugen87 committed:
ObjectLoader: Refactor shape creation

mrdoob commented:
Hmm... It's kind of tempting to move `shapes` to its own group (along `geometries` and `materials` in the JSON format. In order to support this we'll probably need to add a UUID to `Shape` too. Would that make things too complicated?

Mugen87 committed:
Shape: Refactor serialization/ deserialization

Mugen87 commented:
Okay, I've changed the code so shapes are stored in their own group. `ShapeGeometry` and `ShapeBufferGeometry` now refer with a `uuid` to their shapes.

Updated Demo: https://jsfiddle.net/ht3zuzkr/2/

Mugen87 committed:
Clean up

Mugen87 commented:
We could actually also do this for all `Curve` types (so a `Shape` object refers to their curves with `uuid`s) but i have the feeling this goes a little bit too far.

Mugen87 commented:
@mrdoob Is the PR now okay like this? 😊

mrdoob commented:
Checking!

mrdoob commented on src/core/BufferGeometry.js:
```diff
@@ -923,6 +923,32 @@ Object.assign( BufferGeometry.prototype, EventDispatcher.prototype, {
 
+			// ShapeBufferGeometry
+
+			if ( data.type === 'ShapeBufferGeometry' ) {
```
How about adding this in `ShapeBufferGeometry`'s own `toJSON()`?

```js
ShapeBufferGeometry.prototype.toJSON = function () {

	var data = BufferGeometry.prototype.toJSON.call( this );

	data.shapes = [];

	if ( Array.isArray( shapes ) ) {

		for ( var i = 0, l = shapes.length; i < l; i ++ ) {

			var shape = shapes[ i ];

			data.shapes.push( shape.uuid );

		}

	} else {

		data.shapes.push( shapes.uuid );

	}

	return data;

}
```

I think that should work?

Mugen87 replied:
Yes. Indeed, it's better to place this code not in `Geometry` or `BufferGeometry` but in `ShapeGeometry`.

mrdoob commented on src/core/Geometry.js:
```diff
@@ -1009,6 +1009,32 @@ Object.assign( Geometry.prototype, EventDispatcher.prototype, {
 
+			// ShapeGeometry
+
+			if ( data.type === 'ShapeGeometry' ) {
```
Same here...

mrdoob commented:
Apart from these two. It's looking good to me!

Mugen87 committed:
ShapeGeometry: Refacto .toJSON()

Mugen87 commented:
Updated Demo https://jsfiddle.net/ht3zuzkr/2/

mrdoob commented on src/core/Object3D.js:
```diff
@@ -670,6 +671,31 @@ Object.assign( Object3D.prototype, EventDispatcher.prototype, {
+			if ( parameters !== undefined && parameters.shapes !== undefined ) {
+
+				var shapes = parameters.shapes;
+				var uuids = [];
```
What's this array for? 🤔

Mugen87 replied:
Oh, this is a remnant of a previous solution. I'll remove it...

Mugen87 committed:
Object3D: Remove unused array in .toJSON()

mrdoob commented:
Beautiful! 😊

mrdoob commented:
Thanks!
